README.md
[![Release Version][release-shield]][release-url]
[![Downloads][downloads-shield]][downloads-url]
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]

<!-- PROJECT LOGO -->
<br />
<p align="center">
    <img src="readme-assets/facial-recognition.png" alt="Logo" width="150" height="150">

  <h2 align="center">Facial Recognition Dashboard</h2>

  <p align="center">
    A Weekend Project showing off Facial Recognition
    <br />
    <a href="http://ec2-52-66-232-128.ap-south-1.compute.amazonaws.com:5000/">View Demo</a>
    ·
    <a href="https://github.com/animikhaich/Facial-Recognition-Dashboard/issues/new">Report Bug</a>
    ·
    <a href="https://github.com/animikhaich/Facial-Recognition-Dashboard/issues/new">Request Feature</a>
  </p>
</p>

![Facial Recognition Dashboard Homepage][product-screenshot]

<!-- TABLE OF CONTENTS -->

## Table of Contents

- [Table of Contents](#table-of-contents)
- [About The Project](#about-the-project)
  - [Built With](#built-with)
    - [Frontend](#frontend)
    - [Backend](#backend)
    - [Deep Learning](#deep-learning)
- [Getting Started](#getting-started)
  - [Prerequisites](#prerequisites)
  - [Minimum Hardware Requirements](#minimum-hardware-requirements)
  - [Installation](#installation)
- [Usage](#usage)
- [Changelog](#changelog)
- [Roadmap](#roadmap)
- [Contributing](#contributing)
- [License](#license)
- [Contact](#contact)
  - [Animikh Aich](#animikh-aich)
- [Acknowledgements](#acknowledgements)

<!-- ABOUT THE PROJECT -->

## About The Project

A portfolio style dashboard to show Face Detection and Recognition in action. This is also my first attempt at a full-stack deployment of a Deep Learning based project.

It does a few fundamental actions:

- The dashboard provides a UI for the user to interact with the underlying backend
- The pre-trained MTCNN based Face Detection detects faces and prompts the user to label them
- The pre-trained VGG Face extracts the facial features from the cropped faces
- When the user uploads the Query images, Detection and followed by Feature Extraction is done again for the query faces
- Finally, the similarity is between the labeled faces and the query faces are calculated using Cosine Distance Metric and the best match is displayed with a bounding box and name.

### Built With

Below are the languages, libraries and frameworks used for this project

#### Frontend

- [HTML](https://developer.mozilla.org/en-US/docs/Web/HTML)
- [CSS](https://developer.mozilla.org/en-US/docs/Web/CSS)
- [Bootstrap](https://getbootstrap.com)
- [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript)

#### Backend

- [Python](https://www.python.org/)
- [Flask](https://palletsprojects.com/p/flask/)

#### Deep Learning

- [Keras](https://keras.io/)
- [Tensorflow](https://www.tensorflow.org/)

<!-- GETTING STARTED -->

## Getting Started

Running the code to start the web-server is fairly simple. You can easily follow the steps.

### Prerequisites

- [Python 3](https://www.python.org/)
- [Git Command Line](https://git-scm.com/)

### Minimum Hardware Requirements

- CPU: 2 Logical Cores (Threads)
- RAM: 4 GB
- Storage: 10 GB (Including Dependencies)
- OS: Linux, Windows, MacOS
- Internet: 2 Mbps

### Installation

1. Clone the repo

```sh
git clone https://github.com/animikhaich/Facial-Recognition-Dashboard.git
```

2. Install Python packages

```sh
pip install -r requirements.txt
```

3. Run the Flask Server

```sh
python app.py
```

<!-- USAGE EXAMPLES -->

## Usage

This was a weekend project for me. It's primary purpose is to serve as a demo of Facial Recognition. You can easily modify it as per your needs for a college project.

<!-- CHANGELOG -->

## Changelog

See the [Changelog](CHANGELOG.md).

<!-- ROADMAP -->

## Roadmap

See the [open issues](https://github.com/animikhaich/Facial-Recognition-Dashboard/issues?q=is%3Aopen) for a list of proposed features (and known issues).

<!-- CONTRIBUTING -->

## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<!-- LICENSE -->

## License

Distributed under the MIT License. See [LICENSE](LICENSE.md) for more information.

<!-- CONTACT -->

## Contact

#### Animikh Aich

- LinkedIn: [Animikh Aich](https://www.linkedin.com/in/animikh-aich/)
- Email: [animikhaich@gmail.com](mailto:animikhaich@gmail.com)
- Twitter: [@AichAnimikh](https://twitter.com/AichAnimikh)

Live Dashboard: [Facial Recognition dashboard](http://ec2-52-66-232-128.ap-south-1.compute.amazonaws.com:5000/)

<!-- ACKNOWLEDGEMENTS -->

## Acknowledgements

- [ipazc - MTCNN](https://github.com/ipazc/mtcnn)
- [rcmalli - Keras VGGFace](https://github.com/rcmalli/keras-vggface)

<!-- MARKDOWN LINKS & IMAGES -->

[release-shield]: https://img.shields.io/github/release/animikhaich/Facial-Recognition-Dashboard.svg?style=flat-square
[release-url]: https://github.com/animikhaich/Facial-Recognition-Dashboard/releases
[downloads-shield]: https://img.shields.io/github/downloads/animikhaich/Facial-Recognition-Dashboard/total.svg?style=flat-square
[downloads-url]: https://github.com/animikhaich/Facial-Recognition-Dashboard/releases
[contributors-shield]: https://img.shields.io/github/contributors/animikhaich/Facial-Recognition-Dashboard.svg?style=flat-square
[contributors-url]: https://github.com/animikhaich/Facial-Recognition-Dashboard/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/animikhaich/Facial-Recognition-Dashboard.svg?style=flat-square
[forks-url]: https://github.com/animikhaich/Facial-Recognition-Dashboard/network/members
[stars-shield]: https://img.shields.io/github/stars/animikhaich/Facial-Recognition-Dashboard.svg?style=flat-square
[stars-url]: https://github.com/animikhaich/Facial-Recognition-Dashboard/stargazers
[issues-shield]: https://img.shields.io/github/issues/animikhaich/Facial-Recognition-Dashboard.svg?style=flat-square
[issues-url]: https://github.com/animikhaich/Facial-Recognition-Dashboard/issues
[license-shield]: https://img.shields.io/github/license/animikhaich/Facial-Recognition-Dashboard.svg?style=flat-square
[license-url]: https://github.com/animikhaich/Facial-Recognition-Dashboard/blob/master/LICENSE.md
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=flat-square&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/animikh-aich/
[product-screenshot]: readme-assets/project-page.png


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

- Bugfix: Converted PIL image to BGR at each stage for better detection
- Readme Update

## Version 0.1.0 - 2020-04-27

### Added

- First Release
- Custom class for Facial Feature Extraction and Comparison built on top of [Keras VGG-Face](https://github.com/animikhaich/keras-vggface) (Forked from [rcmalli keras-vggface](https://github.com/rcmalli/keras-vggface))
- Modified (to remove OpenCV dependency) [MTCNN face detector](https://github.com/animikhaich/mtcnn) (Forked from [ipazc mtcnn](https://github.com/ipazc/mtcnn))
- Flask Endpoints for each operation as well as to render HTML templates
- HTML Pages:
  - Home
  - About
  - Contact
  - Upload Faces
  - Label Faces
  - Upload Pictures
  - Display Results


test.py
from flask import Flask, redirect, jsonify, request, send_file, url_for, render_template
from utils.utils import *
from PIL import Image, ImageDraw, ImageFont
from datetime import datetime
import numpy as np
import zipfile
import hashlib
import shutil
import sys
import os

# Some Hardcoded Values
ROOT_DATA_DIR = 'static/data'
CROPPED_FACES_DIR = os.path.join(ROOT_DATA_DIR, 'cropped_faces')
RECOGNIZED_FACES_DIR = os.path.join(ROOT_DATA_DIR, 'recognized_faces')

# Flask Config
app = Flask(__name__)
app.config['DEBUG'] = True
"/static/data/cropped_faces/15dae1c1312045f46dcf54df9c9bb00f1.jpg",
"/static/data/cropped_faces/25dae1c1312045f46dcf54df9c9bb00f1.jpg",
"/static/data/cropped_faces/35dae1c1312045f46dcf54df9c9bb00f1.jpg",
"/static/data/cropped_faces/45dae1c1312045f46dcf54df9c9bb00f1.jpg"

content = [
    {
        'id': 0,
        'image': "/static/data/cropped_faces/05dae1c1312045f46dcf54df9c9bb00f1.jpg"
    },
    {
        'id': 1,
        'image': "/static/data/cropped_faces/25dae1c1312045f46dcf54df9c9bb00f1.jpg"
    },
    {
        'id': 2,
        'image': "/static/data/cropped_faces/35dae1c1312045f46dcf54df9c9bb00f1.jpg"
    },
    {
        'id': 3,
        'image': "/static/data/cropped_faces/45dae1c1312045f46dcf54df9c9bb00f1.jpg"
    },
    {
        'id': 4,
        'image': "/static/data/cropped_faces/45dae1c1312045f46dcf54df9c9bb00f1.jpg"
    }
]


@app.route('/', methods=['GET', 'POST'])
@app.route('/home', methods=['GET', 'POST'])
def home():
    if request.method == 'GET':
        return render_template('home.html')

    return redirect(url_for('upload_faces'))


@app.route('/about', methods=['GET'])
def about():
    return render_template('about.html', title="About")


@app.route('/contact', methods=['GET'])
def contact():
    return render_template('contact.html', title="Contact")


@app.route('/upload_faces', methods=['GET', 'POST'])
def upload_faces():
    # Display Page
    if request.method == 'GET':
        return render_template('upload_faces.html', title="Upload")

    # Delete Existing Images
    try:
        shutil.rmtree(CROPPED_FACES_DIR)
    except:
        pass

    # Making sure Folder exists
    if not os.path.isdir(CROPPED_FACES_DIR):
        os.makedirs(CROPPED_FACES_DIR)

    # Get uploaded files
    uploaded_files = request.files.getlist("file")
    face_num = 0

    faces_list = [uploaded_file.filename for uploaded_file in uploaded_files]
    return jsonify(faces_list)


@app.route('/label_faces', methods=['POST', 'GET'])
def label_faces():
    # Display Page
    if request.method == 'GET':
        return render_template('label_faces.html', title="Label", table_contents=content)

    face_list = list()
    label_list = list()
    for element in content:
        face_id = element.get('id')
        image_path = element.get('image')
        label = request.form.get(f"face-name-{face_id}")
        print(face_id, image_path, label)

        if not label:
            continue

        # image = Image.open(image_path[1:]).convert('RGB')

        # Create the List to pass to FR Module
        face_list.append(image_path)
        label_list.append(label)

    return jsonify([face_list, label_list])


@app.route('/upload_pictures', methods=['POST', 'GET'])
def upload_pictures():

    # Display Page
    if request.method == 'GET':
        return render_template('upload_pictures.html', title="Upload")

    # Delete Existing Images
    try:
        shutil.rmtree(RECOGNIZED_FACES_DIR)
    except:
        pass

    # Making sure Folder exists
    if not os.path.isdir(RECOGNIZED_FACES_DIR):
        os.makedirs(RECOGNIZED_FACES_DIR)

    # Get uploaded files
    uploaded_files = request.files.getlist("file")

    # Detect Faces for each Uploaded Image
    recognized_image_list = [
        uploaded_file.filename for uploaded_file in uploaded_files]

    return jsonify(recognized_image_list)


if __name__ == '__main__':
    app.run(
        host='0.0.0.0',
        port='5000',
        use_reloader=True,
        threaded=True
    )


main.py
from flask import Flask, redirect, jsonify, request, send_file, url_for, render_template
from keras.backend.tensorflow_backend import set_session
from .mtcnn import MTCNN
from utils.utils import *
from utils.vgg_face import VGGFaceRecognizer
from PIL import Image, ImageDraw, ImageFont
import tensorflow as tf
from datetime import datetime
import numpy as np
import zipfile
import hashlib
import shutil
import sys
import os
import time

# Configure Environment for GPU
os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'

sess = tf.Session()
graph = tf.get_default_graph()
set_session(sess)

# Some Hardcoded Values
ROOT_DATA_DIR = 'static/data'
CROPPED_FACES_DIR = os.path.join(ROOT_DATA_DIR, 'cropped_faces')
RECOGNIZED_FACES_DIR = os.path.join(ROOT_DATA_DIR, 'recognized_faces')
mtcnn = MTCNN()
face_recognizer = VGGFaceRecognizer(model='senet50')

# Flask Config
app = Flask(__name__)
app.config['DEBUG'] = False

# In App Variables
content = list()


@app.route('/', methods=['GET', 'POST'])
@app.route('/home', methods=['GET', 'POST'])
def home():
    if request.method == 'GET':
        return render_template('home.html')

    return redirect(url_for('upload_faces'))


@app.route('/about', methods=['GET'])
def about():
    return render_template('about.html', title="About")


@app.route('/contact', methods=['GET'])
def contact():
    return render_template('contact.html', title="Contact")


@app.route('/upload_faces', methods=['POST', 'GET'])
def upload_faces():
    global content, graph
    # Display Page
    if request.method == 'GET':
        return render_template('upload_faces.html')

    # Delete Existing Images
    try:
        shutil.rmtree(CROPPED_FACES_DIR)
    except:
        pass

    # Making sure Folder exists
    if not os.path.isdir(CROPPED_FACES_DIR):
        os.makedirs(CROPPED_FACES_DIR)

    # Get uploaded files
    uploaded_files = request.files.getlist("file")

    face_id = 0

    # Detect Faces for each Uploaded Image
    content = list()
    for uploaded_file in uploaded_files:
        image = Image.open(uploaded_file).convert('RGB')
        image = np.asarray(image)

        # Detect Faces using MTCNN
        with graph.as_default():
            set_session(sess)
            detected_faces = mtcnn.detect_faces(image)

        height, width, _ = image.shape
        for detected_face in detected_faces:
            x1, y1, x2, y2 = fix_coordinates(
                detected_face['box'], width, height
            )
            cropped_face = image[y1:y2, x1:x2]
            try:
                cropped_face = Image.fromarray(cropped_face)
            except:
                continue
            saved_image_path = os.path.join(
                CROPPED_FACES_DIR,
                rem_punctuation(f"{face_id}_{uploaded_file.filename}")
            )
            cropped_face.save(saved_image_path)
            content.append(
                {
                    'id': face_id,
                    'image': f"/{saved_image_path}"
                }
            )
            face_id += 1

    return redirect(url_for('label_faces'))


@app.route('/label_faces', methods=['POST', 'GET'])
def label_faces():
    global content, graph

    # Display Page
    if request.method == 'GET':
        return render_template('label_faces.html', title="Label", table_contents=content)

    face_list = list()
    label_list = list()
    for element in content:
        face_id = element.get('id')
        image_path = element.get('image')
        label = request.form.get(f"face-name-{face_id}")
        print(face_id, image_path, label)

        if not label:
            continue

        # Create the List to pass to FR Module
        image = Image.open(image_path[1:]).convert('RGB')
        face_list.append(image)
        label_list.append(label)

    # Register Faces - Extract and store ground truth Facial Features
    with graph.as_default():
        set_session(sess)
        face_recognizer.register_faces(face_list, label_list)

    return redirect(url_for('upload_pictures'))


@app.route('/upload_pictures', methods=['POST', 'GET'])
def upload_pictures():
    global graph

    # Display Page
    if request.method == 'GET':
        return render_template('upload_pictures.html', title="Upload")

    # Delete Existing Images
    try:
        shutil.rmtree(RECOGNIZED_FACES_DIR)
    except:
        pass

    # Making sure Folder exists
    if not os.path.isdir(RECOGNIZED_FACES_DIR):
        os.makedirs(RECOGNIZED_FACES_DIR)

    # Get uploaded files
    uploaded_files = request.files.getlist("file")

    # Detect Faces for each Uploaded Image
    recognized_image_list = list()
    for uploaded_file in uploaded_files:
        image_PIL = Image.open(uploaded_file).convert('RGB')
        image = np.asarray(image_PIL)

        # Detect Faces using MTCNN
        with graph.as_default():
            set_session(sess)
            detected_faces = mtcnn.detect_faces(image)

        height, width, _ = image.shape
        font_size = int(np.mean([width, height]) * 0.03)

        # TODO: From here down
        draw = ImageDraw.Draw(image_PIL)
        for detected_face in detected_faces:
            x1, y1, x2, y2 = fix_coordinates(
                detected_face['box'], width, height
            )
            cropped_face = image[y1:y2, x1:x2]
            try:
                cropped_face = Image.fromarray(cropped_face)
            except:
                continue

            with graph.as_default():
                set_session(sess)
                face_name = face_recognizer.recognize(
                    cropped_face, thresh=0.25)
            if face_name:
                draw.rectangle((x1, y1, x2, y2))
                font = ImageFont.truetype(
                    "static/assets/fonts/Roboto-Regular.ttf", font_size)

                text_w, text_h = draw.textsize(face_name)
                text_x = int(x1)
                text_y = int(y1 - font_size)
                draw.text((text_x, text_y), face_name, fill='red', font=font)

        image_path = os.path.join(
            RECOGNIZED_FACES_DIR, rem_punctuation(
                f"{time.time()}_{uploaded_file.filename}")
        )
        image_PIL.save(image_path)
        recognized_image_list.append('/'+image_path)

    return redirect(url_for('display_results'))


@app.route('/display_results', methods=['GET'])
def display_results():
    # Get List of images in Recognized Faces Directory
    images = os.listdir(RECOGNIZED_FACES_DIR)
    image_paths = ['/'+os.path.join(RECOGNIZED_FACES_DIR, im_path).replace(' ', '%20')
                   for im_path in images]

    return render_template('display_results.html', image_paths=image_paths)


if __name__ == '__main__':
    app.run(
        host='0.0.0.0',
        port='5000',
        use_reloader=False,
        threaded=True
    )


utils/vgg_face.py
from keras.backend.tensorflow_backend import set_session
from keras.preprocessing.image import load_img
from keras_vggface.utils import preprocess_input
from keras_vggface.vggface import VGGFace
from scipy.spatial.distance import cosine
import tensorflow as tf
from PIL import Image
import numpy as np
import os

# Set environment variable to make sure that TF does not eat up all GPU memory
os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'


class VGGFaceRecognizer:
    """
    Easy Face Recognizer Class, built on top of RC Mali's VGG Face Implementation
    URL (Keras VGG Face): https://github.com/rcmalli/keras-vggface
    License: MIT

    Author: Animikh Aich
    GitHub: https://github.com/animikhaich
    LinkedIn: https://www.linkedin.com/in/animikh-aich/
    """

    def __init__(self, model='senet50'):
        """
        __init__ Model Initializer

        - Sets the flags for GPU memory growth (if available)
        - Selects and Initializes the Model

        Args:
            model (str, optional): Choose Base Feature Extractor: VGG16, ResNet50, and SeNet50. Defaults to 'senet50'.
        """

        # Allow Memory Growth for Tensorflow GPU
        if tf.test.is_gpu_available():
            # For TF 1x
            if int(tf.__version__.split('.')[0]) == 1:
                config = tf.ConfigProto()
                config.gpu_options.allow_growth = True
                session = tf.Session(config=config)
                set_session(session)

            # For TF 2x
            else:
                gpu_devices = tf.config.experimental.list_physical_devices(
                    'GPU'
                )
                tf.config.experimental.set_memory_growth(gpu_devices[0], True)

        # Some Instance Variables
        self.model = model
        self.face_dict = dict()

        # initialize VGG Face using VGGFace Model
        self.recognizer = VGGFace(include_top=False, model=model)

    def calculate_similarity(self, vector_1, vector_2):
        """
        calculate_similarity

        Calculate the Cosine Similarity of Similarity Feature Vectors
        Flattens the Vectors for VGG16 base due to higher high dimension vector

        Args:
            vector_1 (numpy array): Feature Vector of Face 1
            vector_2 (numpy array): Feature Vector of Face 2

        Returns:
            float: Similarity Score, Lower means more similar
        """
        # VGG creates higher dimensional vector, and hence needs to be flattened
        if self.model == 'vgg16':
            vector_1 = vector_1.flatten()
            vector_2 = vector_2.flatten()

        # Remove Extra Dimension
        vector_1 = np.squeeze(vector_1)
        vector_2 = np.squeeze(vector_2)

        # Calculate the cosine similarity
        return cosine(vector_1, vector_2)

    def register_faces(self, face_list, name_list):
        """
        register_faces 

        Given a list of faces and names, register them and save them to the database
        These are the faces that a query face will be compared against.
        Every time this is called, it clear the existing database

        Args:
            face_list (list): List of PIL Face Images
            name_list (list): List of Strings containing the name of each face - Unique

        Returns:
            dict: Dictionary containing the name of the face and it's corresponding feature vector
        """
        self.face_dict = dict()
        for face, name in zip(face_list, name_list):
            self.face_dict[name] = self.feature_extractor(face)

        return self.face_dict

    def add_faces(self, face_list, name_list):
        """
        add_faces 

        Given a list of faces and names, register them and save them to the database
        These are the faces that a query face will be compared against.
        Every time this is called, it appends to existing faces in the DB

        Args:
            face_list (list): List of PIL Face Images
            name_list (list): List of Strings containing the name of each face - Unique

        Returns:
            dict: Dictionary containing the name of the face and it's corresponding feature vector
        """
        for face, name in zip(face_list, name_list):
            self.face_dict[name] = self.feature_extractor(face)

        return self.face_dict

    def feature_extractor(self, face):
        """
        feature_extractor 

        Primary CNN based feature extractor for faces. Uses pre-trained model by RC Mali for the same.
        You can choose the architecture while creating the instance of the class

        Args:
            face (PIL Image): PIL image of a cropped out face

        Returns:
            list: Returns the feature vector that is extracted from the given face
        """
        # Resize the face, face is a PIL image
        face = face.resize((224, 224), Image.ANTIALIAS)
        face = np.asarray(face).astype(np.float)
        face = np.expand_dims(face, axis=0)

        if self.model == 'vgg16':
            face = preprocess_input(face, version=1)
        else:
            face = preprocess_input(face, version=2)

        return self.recognizer.predict(face)

    def recognize(self, face, thresh=0.25):
        """
        recognize 

        Given a Query Face, this function compares it against the existing faces and returns a match, if any

        Args:
            face (PIL Image): Cropped face in the form of a PIL image
            thresh (float, optional): Cutoff Threshold used to discard unmatched faces. Defaults to 0.25.

        Returns:
            string: Name of the matched face, if recognized, else None
        """
        # Extract features of the Query Face
        query_features = self.feature_extractor(face)
        temp_sim_dict = dict()

        # Compare the Query Face features with the existing database faces
        for key in self.face_dict.keys():
            db_face_features = self.face_dict[key]
            score = self.calculate_similarity(
                db_face_features, query_features
            )
            temp_sim_dict[key] = score

        try:
            # Return None if no face matches
            if min(temp_sim_dict.values()) > thresh:
                return None
        except:
            return None

        # Find most similar face
        most_similar_face = min(temp_sim_dict, key=temp_sim_dict.get)

        return most_similar_face


app.py
from flask import Flask, redirect, jsonify, request, send_file, url_for, render_template
from keras.backend.tensorflow_backend import set_session
from mtcnn import MTCNN
from utils.utils import *
from utils.vgg_face import VGGFaceRecognizer
from PIL import Image, ImageDraw, ImageFont
import tensorflow as tf
from datetime import datetime
import numpy as np
import zipfile
import hashlib
import shutil
import sys
import os
import time

# Configure Environment for GPU
os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'

sess = tf.Session()
graph = tf.get_default_graph()
set_session(sess)

# Some Hardcoded Values
ROOT_DATA_DIR = 'static/data'
CROPPED_FACES_DIR = os.path.join(ROOT_DATA_DIR, 'cropped_faces')
RECOGNIZED_FACES_DIR = os.path.join(ROOT_DATA_DIR, 'recognized_faces')
mtcnn = MTCNN()
face_recognizer = VGGFaceRecognizer(model='senet50')

# Flask Config
app = Flask(__name__)
app.config['DEBUG'] = False

# In App Variables
content = list()


@app.route('/', methods=['GET', 'POST'])
@app.route('/home', methods=['GET', 'POST'])
def home():
    if request.method == 'GET':
        return render_template('home.html')

    return redirect(url_for('upload_faces'))


@app.route('/about', methods=['GET'])
def about():
    return render_template('about.html', title="About")


@app.route('/contact', methods=['GET'])
def contact():
    return render_template('contact.html', title="Contact")


@app.route('/upload_faces', methods=['POST', 'GET'])
def upload_faces():
    global content, graph
    # Display Page
    if request.method == 'GET':
        return render_template('upload_faces.html')

    # Delete Existing Images
    try:
        shutil.rmtree(CROPPED_FACES_DIR)
    except:
        pass

    # Making sure Folder exists
    if not os.path.isdir(CROPPED_FACES_DIR):
        os.makedirs(CROPPED_FACES_DIR)

    # Get uploaded files
    uploaded_files = request.files.getlist("file")

    face_id = 0

    # Detect Faces for each Uploaded Image
    content = list()
    for uploaded_file in uploaded_files:
        image = Image.open(uploaded_file).convert('RGB')
        image = np.asarray(image)
        image = image[:, :, ::-1]  # Convert RGB to BGR

        # Detect Faces using MTCNN
        with graph.as_default():
            set_session(sess)
            detected_faces = mtcnn.detect_faces(image)
            print(detected_faces)

        height, width, _ = image.shape
        for detected_face in detected_faces:
            x1, y1, x2, y2 = fix_coordinates(
                detected_face['box'], width, height
            )
            cropped_face = image[y1:y2, x1:x2]
            try:
                cropped_face = Image.fromarray(cropped_face)
            except:
                continue
            saved_image_path = os.path.join(
                CROPPED_FACES_DIR,
                rem_punctuation(f"{face_id}_{uploaded_file.filename}")
            )
            cropped_face.save(saved_image_path)
            content.append(
                {
                    'id': face_id,
                    'image': f"/{saved_image_path}"
                }
            )
            face_id += 1

    return redirect(url_for('label_faces'))


@app.route('/label_faces', methods=['POST', 'GET'])
def label_faces():
    global content, graph

    # Display Page
    if request.method == 'GET':
        return render_template('label_faces.html', title="Label", table_contents=content)

    face_list = list()
    label_list = list()
    for element in content:
        face_id = element.get('id')
        image_path = element.get('image')
        label = request.form.get(f"face-name-{face_id}")
        print(face_id, image_path, label)

        if not label:
            continue

        # Create the List to pass to FR Module
        image = Image.open(image_path[1:]).convert('RGB')
        face_list.append(image)
        label_list.append(label)

    # Register Faces - Extract and store ground truth Facial Features
    with graph.as_default():
        set_session(sess)
        face_recognizer.register_faces(face_list, label_list)

    return redirect(url_for('upload_pictures'))


@app.route('/upload_pictures', methods=['POST', 'GET'])
def upload_pictures():
    global graph

    # Display Page
    if request.method == 'GET':
        return render_template('upload_pictures.html', title="Upload")

    # Delete Existing Images
    try:
        shutil.rmtree(RECOGNIZED_FACES_DIR)
    except:
        pass

    # Making sure Folder exists
    if not os.path.isdir(RECOGNIZED_FACES_DIR):
        os.makedirs(RECOGNIZED_FACES_DIR)

    # Get uploaded files
    uploaded_files = request.files.getlist("file")

    # Detect Faces for each Uploaded Image
    recognized_image_list = list()
    for uploaded_file in uploaded_files:
        image_PIL = Image.open(uploaded_file).convert('RGB')
        image = np.asarray(image_PIL)
        image = image[:, :, ::-1]  # Convert RGB to BGR

        # Detect Faces using MTCNN
        with graph.as_default():
            set_session(sess)
            detected_faces = mtcnn.detect_faces(image)

        height, width, _ = image.shape
        font_size = int(np.mean([width, height]) * 0.03)

        draw = ImageDraw.Draw(image_PIL)
        for detected_face in detected_faces:
            x1, y1, x2, y2 = fix_coordinates(
                detected_face['box'], width, height
            )
            cropped_face = image[y1:y2, x1:x2]
            try:
                cropped_face = Image.fromarray(cropped_face)
            except:
                continue

            with graph.as_default():
                set_session(sess)
                face_name = face_recognizer.recognize(
                    cropped_face, thresh=0.30)
            if face_name:
                draw.rectangle((x1, y1, x2, y2), width=5)
                font = ImageFont.truetype(
                    "static/assets/fonts/Roboto-Regular.ttf", font_size)

                text_w, text_h = draw.textsize(face_name)
                text_x = int(x1)
                text_y = int(y1 - font_size)
                draw.text((text_x, text_y), face_name, fill='red', font=font)

        image_path = os.path.join(
            RECOGNIZED_FACES_DIR, rem_punctuation(
                f"{time.time()}_{uploaded_file.filename}")
        )
        image_PIL.save(image_path)
        recognized_image_list.append('/'+image_path)

    return redirect(url_for('display_results'))


@app.route('/display_results', methods=['GET'])
def display_results():
    # Get List of images in Recognized Faces Directory
    images = os.listdir(RECOGNIZED_FACES_DIR)
    image_paths = ['/'+os.path.join(RECOGNIZED_FACES_DIR, im_path).replace(' ', '%20')
                   for im_path in images]

    return render_template('display_results.html', image_paths=image_paths)


if __name__ == '__main__':
    app.run(
        host='0.0.0.0',
        port='5000',
        use_reloader=False,
        threaded=True
    )
